sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -189,7 +189,14 @@ public int read(ByteBuffer dst) throws IOException {
       if (closed) {
         throw new IOException("Channel is closed");
       }
-      return inputStream.read(dst);
+      // We avoid using the ByteBuffer based read for Hadoop because some FSInputStream
+      // implementations are not ByteBufferReadable,
+      // See https://issues.apache.org/jira/browse/HADOOP-14603
+      int read = inputStream.read(dst.array());
```

Comment:
```dst``` make have a limit on how much to read and its start position may not be zero.
```dst``` may not have array accessible because the bytes may be off heap.

You should use position and limit to write into the bytebuffer or you will not respect what the user has asked for.
You should only use ```array()``` if you checked ```hasArray()``` and correctly calculate the offsets/limits with ```position()```, ```arrayOffset()``` and ```remaining```. Otherwise you should use a fixed size buffer and copy the bytes twice.